compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/cfa/PropertyInitializationInfoCache.kt

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright 2010-2021 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.fir.analysis.cfa
+
+import org.jetbrains.kotlin.fir.resolve.dfa.cfg.CFGNode
+import org.jetbrains.kotlin.fir.resolve.dfa.cfg.ControlFlowGraph
+import org.jetbrains.kotlin.fir.symbols.impl.FirPropertySymbol
+
+/**
+ * A Cache that maintains mapping from CFG to collected property initialization info.
+ * [FirMemberPropertiesChekcer] will analyze constructors, anonymous initializers, and property initializers to determine if member
+ * properties are initialized. To avoid redundant control-flow analysis, that checker will collect information for member properties
+ * as well as local properties. Then, when [FirControlFlowAnalyzer] wants to analyze those functions again, the cache will be used.
+ *
+ * But, if we cache everything, the mappings may grow too much. For regular uses (from [FirControlFlowAnalyzer]), we don't need caching.
+ * We will do so only for constructors, anonymous initializers, and property initializers visited ahead by [FirMemberPropertiesChecker].
+ */
+class PropertyInitializationInfoCache private constructor() {
+    // TODO: we may need a way to invalidate some cache data, e.g., if the containing declaration is edited in IDE.
```

Comment:
IIUC the IDE currently re-compute the FIR tree from ground up on use edits. Are there plans to add incremental updating FIR?

Reply:
> Are there plans to add incremental updating FIR?

Not sure about that.

Good to know that the FIR tree will be rebuilt from scratch, which means the stale cache won't reused anyway. Actually, it's not a big deal to add an API to invalidate cache, so I added one.